### ISSUE TITLE
Publish API/button needs to send message(s) to external endpoint on publication

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#3561]
As discussed with @pameyer @djbrooke and @scolapasta [yesterday](https://docs.google.com/document/d/1Cp0myGJKAMWQLIT8wHE8JLitUOsySUM0S7yQ_rx4Cgw/edit?usp=sharing) the rsync solution we are working on in #3145 requires potentially time intensive movement of files from one filesystem called "hold" to another filesystem called "public" just before the dataset is published in Dataverse. In the [doc](https://docs.google.com/a/hkl.hms.harvard.edu/document/d/1idgIT_BOOGDuhR5j9RSOi-3NdrPhRS60N2O953LvzVM/edit?usp=sharing) we reviewed we called this "Publish API/button needs to send message(s) to external endpoint on publication" and the external endpoint in question is currently known as "The Backend" but will eventually be given an name and will expose a JSON-based REST endpoint. The communication between Dataverse and The Backend will go something like this:

- Dataverse tells The Backend to move files from "hold" to "public"
  - Note that there are UI/UX considerations for this workflow in the GUI. Perhaps the user still clicks "Publish" but an indication needs to be shown that publishing will take a long time, perhaps overnight, due to the size of the dataset or simply because files were uploaded via an rsync script. See also #3348.
- The Backend replies immediately with some sort of job id confirming that the message was recieved.
- Time passes. Files are moved by The Backend.
- The Backend tells Dataverse that file movement is complete.
- Dataverse publishes the dataset.

We will iterate on this description and provide more details as needed.

**Comment by pameyer** [comment]
"the backend" endpoint for Dataverse to communicate with needs to be defined/implemented.  I'll set up another repo (analogous to how we approached the DCM) for that component; (initially with mocks) so that there's something to develop against.

**Comment by pameyer** [comment]
Until renamed (and to avoid namespace confusion), "the backend" is [RSAL](https://github.com/sbgrid/rsal).

Very preliminary API docs there, mocks to follow.

**Comment by pameyer** [comment]
and mocks are in that repo as well

**Comment by djbrooke** [comment]
We discussed this in sprint planning today, and @pameyer mentioned that @michbarsinai  currently has what he needs to move this forward and will continue to provide things as needed.

**Comment by pdurbin** [comment]
It has been requested that the "Design Options for Remotely Managed Datasets" doc by @michbarsinai be reviewed by all interested parties in time to discuss it during the regular Tuesday morning call next week (Feb 14): https://docs.google.com/document/d/1_m4wHYsocNGDj-uBmBCXjwh0kVPyS5GyThjKnigS1zg/edit?usp=sharing

**Comment by pameyer** [comment]
"Design Options for Remotely Managed Datasets" not currently applicable (to this issue), since @michbarsinai 's redesign (and subsequent meetings/conversations about it).

Some of these discussions touched on the need to prevent users from changing a dataset (or initiating other state transitions) at various points of the process; this intersects with the idea of ensuring that datasets aren't changeable while the dataset is being reviewed (#3172).

**Comment by pdurbin** [comment]
@pameyer good point. Here's a link to the new "Design Document for Publication Workflows" doc that we've switch to: https://docs.google.com/document/d/1GFrCh23TdYc87dBi4xg5Pl2_Gf1usQqrqaik2-i7V7A/edit?usp=sharing

**Comment by mheppler** [comment]
After reviewing the aforementioned [design doc](https://docs.google.com/document/d/1GFrCh23TdYc87dBi4xg5Pl2_Gf1usQqrqaik2-i7V7A/edit) in our mtg yesterday, we agreed that for now, we'll be able to utilize the UX example of dataset locking as it exists for tabular data file ingest.

Taking that example of disabling action buttons, we'll also use the message block at the top of the page to display a warning, with more details, as well as User Guide and Contact Support links. There is also the example of "Draft", "Unpublished" and "In Review" publication status labels next to the dataset title, where we could add new labels for "Reviewed" or any other state we may need to display.

![mockup-dataset-lock](https://cloud.githubusercontent.com/assets/687227/23526980/1343f29e-ff62-11e6-9b3c-37495b873d54.jpg)

**Comment by pdurbin** [comment]
I just mentioned this issue to @donsizemore at http://irclog.iq.harvard.edu/dataverse/2017-05-16#i_52743 because his problem reminds of the reason we are introducing this "workflows" concept. The operation his user is attempting is slow:

> thu-mai prepares to publish a dataset with 224 files in it. gears whir, white-gloved hands with knives start spinning. asynchronously, backgrounded. in the browser, thu-mai gets the spinning wheel of work
> 
> she waits a while, tries to publish again, gets an error. i'm at lunch. ingest and indexing can take a while with large datasets, but any way to get good/meaningful feedback to the user until it completes? publishing-in-progress semaphore or some such?

Also, I'd like to mention that @michbarsinai has been working away on a branch at https://github.com/IQSS/dataverse/tree/3561-publication-workflows that's ready for some code review.

**Comment by pdurbin** [comment]
Yesterday I asked @djbrooke about the status of https://github.com/IQSS/dataverse/tree/3561-publication-workflows and he indicated that it is not ready for code review. @michbarsinai is still working away on it. My bad.

I just left a related comment on this effort at https://github.com/IQSS/dataverse/issues/3725#issuecomment-302399257 that might be of interest.

**Comment by pameyer** [comment]
at least according to my notes; @michbarsinai mentioned this was ready for an initial (non-UI) code review 5/2.

**Comment by pdurbin** [comment]
@pameyer I dunno. Yesterday @michbarsinai mentioned `ServiceLoader` and `NoClassFoundException`'s so I think he's still plugging away. Sounds like SPI stuff (#3657).

**Comment by michbarsinai** [comment]
The parts ready for the code review are the new DatasetPublishCommand, which was heavily refactored.

**Comment by michbarsinai** [comment]
Some SPI pointers in @pdurbin 's comment:
https://github.com/IQSS/dataverse/issues/3657#issuecomment-302432091

**Comment by pameyer** [comment]
Possibly worth mentioning in the documentation that calls to `api/workflows/$invokation_id` need to have the content-type header set (aka - this info should live somewhere other than my notes).

**Comment by pameyer** [comment]
First pass run-through:
- [ ] batch-import component of dcm success API triggers exception when trying to lock.
[server-dcm_ok01_edit.txt](https://github.com/IQSS/dataverse/files/1264886/server-dcm_ok01_edit.txt). success message goes out; package file rename to storage identifier doesn't happen, file doesn't show up on dataset page (success message going out on failure probably fixed in https://github.com/IQSS/dataverse/issues/3348).  Investigating.

**Comment by sekmiller** [comment]
The inReview Lock needs to be ignored when the user has permission to publish a dataset. ( @pameyer notes that above)
Also, there's a timing issue with the creation of inReview locks via the API where the response isn't seeing the lock so the inReviewWorkflow integration test fails.  I thought there might be a quick fix for this, but I haven't found it yet.

**Comment by landreev** [comment]
@pameyer - this item on your report from last night: "Dataset gets stuck, unable to publish via either UI or API." The fix that @sekmiller made appears to have fixed the UI part... Just to confirm,  the API part of the issue is still an issue, correct?

**Comment by sekmiller** [comment]
@pameyer It is. I am moving onto the API piece now.

**Comment by landreev** [comment]
@sekmiller - ok, please let me know if you want me to take a look too. 

@pameyer - the first issue ("exception when trying to lock") is probably not an issue, because currently in the branch 3348-RSync-FileUpload-Workflow we are no longer locking the dataset in that part of the code; this is because it's now getting locked the moment the user downloads the script. So, should I comment out this line? - we may need to revisit this, once we merge this branch into 3348. do we need to do something more intelligent there? Check that the dataset is indeed locked, waiting for an rsync upload to happen? Should the crawler job remove that lock, and create a new, specifically marked to indicate that the crawler job is already in progress - thus preventing another one to happen in parallel? But I think it makes more sense to just comment out the line in this PR, and move ahead and address this in 3348, if necessary.

**Comment by landreev** [comment]
@pameyer oh, and the false success notification has been fixed in #3348, yes.

**Comment by sekmiller** [comment]
@landreev In the Return to Author Command I just tried changing the dataset passed to the remove lock command to the updated dataset. The integration tests are still failing because the lock still seems to be in place. Will try to figure out why.

**Comment by landreev** [comment]
@pameyer @michbarsinai should we add the 4th type of lock - specifically for rsync uploads in progress? "RSYNC_UPLOAD" or "DCM_UPLOAD"? 
We want to be able to differentiate between a tabular ingest in progress, and a dcm upload in progress. Because the page needs to show different messages (and maybe more). 

In the current #3348 implementation I'm achieving this by looking at the Info field in the lock. If it says "DCM ..." I'm assuming it's a DCM job in progress... So I can continue doing this (this would actually be easier - less code to change...) But it could be prettier to have a dedicated lock for this - ? 

Not super attached to either way... let me know what you think please.

**Comment by landreev** [comment]
@sekmiller - I think it's because RemoveLockCommand just calls the DatasetServiceBean to remove the lock, but does not update the Dataset object passed to it. Just adding a 
dataset.setDatasetLock(null); 
line to it should fix it, I think?

**Comment by pameyer** [comment]
@sekmiller @landreev That looks like the same review stuff to me.

"exception while trying to lock" - Commenting it out and revisiting in 3348 sounds good to me.

"add 4th type of lock" - adding a lock type for a new type of lock seems reasonable to me; but no strong preference either way. @michbarsinai / @scolapasta may have better insight.

**Comment by landreev** [comment]
@michbarsinai @pameyer 

This came up during the standup today: 

The plan appears to be to subject this branch to the normal QA process before it's merged into develop (correct?). So we need to let @kcondon know what exactly to test. My understanding of it: 

- No QA of any actual rsync functionality (that will be done when the 2 rsync-specific branches are ready);
- Need to QA for regression on anything in the existing functionality related to locks and the "in review" dataset state; because these things have been changed in this PR. 
- The only place where Dataset locks are currently being used is tabular Ingest. So testing should involve confirming that the behavior hasn't changed - that starting tab ingests from the dataset page locks the dataset; that the page stays locked until the ingest is done; and that the dataset gets unlocked when it finishes. 
- Submitting the dataset for review should continue working as it has been before. (I realize I don't know much about this part...)
- The NEW piece of functionality for which the issue was opened originally: the whole thing about messaging the external system during the publishing workflow. (Pete confirmed last night that this appears to be working - see his comment above - but Kevin may need instructions on how to actually test it with RSAL - ?)

Anything I'm missing? 

Would be nice to merge this thing soon!

**Comment by pameyer** [comment]
- "No QA of any actual rsync functionality"; there's some exiting functionality that's already been merged into develop, but if it's more efficient I think we can defer those into the other two branches.
- "Need to QA for regression"/"in review" and ~"confirm ingest behavior"; I think you're correct on both.
- "Submitting dataset for review"; this should either behave as before or address #3172 (datasets submitted for review not being editable by the submitter). I'd be in favor of the second.
- "new functionality"; there's an RSAL mock that can be used for this (analogous to the DCM mock) that I can set @kcondon up with to help test that part.

**Comment by landreev** [comment]
"there's some exiting functionality that's already been merged into develop" - if it's already been merged, then it must have been already QAed too. So you're talking about re-testing it, to make sure it didn't get broken by the changes in this PR, correct? 
I don't know. Not sure it makes sense, to re-test batch import (for example), as part of merging this PR... Considering that we'll be QAing it again, as part of #3348.

**Comment by pameyer** [comment]
That was what I was talking about; if it makes more sense to wait until #3348 / #3998, then let's do it that way.

**Comment by michbarsinai** [comment]
In addition to the existing locks listed by @landreev, need to check the workflows themselves:

1/ Managing workflows using the management API
2/ Setting and un-setting Pre- and Post- release workflows 
3/ Whitelisting IP addresses
4/ Sending outside messages, waiting for external system to send POST request (this is covered by testing the RSAL functionality)

**Comment by sekmiller** [comment]
I fixed the Publish command to allow an InReview Dataset to be published. I think we also need to return Edit Dataset Privileges on InReview Datasets to Curators

**Comment by landreev** [comment]
Note from the group review: Show the info message associated with the lock on the dataset page; for all types of locks. (currently, only the "dcm upload in progress" message is shown). 
Should be easy to do - but need to run it by @mheppler.

**Comment by landreev** [comment]
"This dataset has been published" success message appears to be automatically tied to the user hitting the 
"publish" button, and not getting an exception. We don't want to show it if there is a workflow in progress!

**Comment by landreev** [comment]
The "unpublished" tag immediately disappears, even when there is a workflow in progress.

**Comment by pameyer** [comment]
failure case when workflow uses an external system, and the external system is down.  results in an unpublished dataset w\ the published button greyed out; probably should report an error and clear the workflow lock.

**Comment by landreev** [comment]
I added an info message for when the dataset is locked in the publish workflow; the actual message, dataset.publish.workflow.inprogress in the bundle is just a placeholder:

Publish workflow in progress

please replace it with something sensible

**Comment by landreev** [comment]
I fixed the missing "unpublished" tag by moving the setPublicationDate() line  from the PublishDatasetCommand to FinalizePublishDatasetCommand (publication date being non null is the definition used by dvobject.isReleased(), used in turn by the dataset page to determine if the dataset is published). 
I'm not 100% sure if this isn't going to break something somewhere else. And I'd like to quickly review the overall functionality split between PublishDatasetCommand and FinalizePublishDatasetCommand with somebody else... @scolapasta ?

**Comment by landreev** [comment]
If the fixes I made today are legit, this should be ready for QA. 
but let's review tomorrow.

**Comment by michbarsinai** [comment]
@pmeyer - this should have been taken care of by the catch clauses here:

https://github.com/IQSS/dataverse/blob/3561-publication-workflows/src/main/java/edu/harvard/iq/dataverse/workflow/internalspi/HttpSendReceiveClientStep.java#L40 <https://github.com/IQSS/dataverse/blob/3561-publication-workflows/src/main/java/edu/harvard/iq/dataverse/workflow/internalspi/HttpSendReceiveClientStep.java#L40>

Reporting error is done using returning a Failure object, which causes the rollback system to roll-back the successful steps done so far.

**Comment by pameyer** [comment]
@michbarsinai I'll re-test; but it was looking like there was a lock that wasn't getting cleared on rollback.  I'll update with more definite information once I've done some investigation.

**Comment by pameyer** [comment]
More investigation:
- create external workflow with `http/sr` stepType for pre-publish
- create dataset
- submit for review
- turn off the system listening at the `url` for the `http/sr` step.
- admin user published dataset (through UI)
- server log shows workflow engine rolling back workflow; dataset page shows (new) message about publication workflow in progress, de-activated publish button, no return to author button.

Same behavior seen for publishing without a submit for review step; also seen with publishing through API.

**Comment by pameyer** [comment]
From discussion at stand-up and some investigation, this is probably better off in its own issue (source of the problems is system misconfiguration/failure, and there's a relatively straightforward way to get out of the stuck state (for a sysadmin).

**Comment by landreev** [comment]
@michbarsinai 
@scolapasta and I have just reviewed the PublishDatasetCommand vs. FinalizeDatasetPublicationCommand functionality split and we have the following question: 

What there a specific reason why the following steps were left in PublishDatasetCommand: 

```
             theDataset.setPublicationDate(new Timestamp(new Date().getTime()));
             theDataset.getEditVersion().setVersionNumber(new Long(1)); 
            theDataset.getEditVersion().setMinorVersionNumber(new Long(0));
 (for a previously unpublished dataset)
            ...
            Timestamp updateTime = new Timestamp(new Date().getTime());
            theDataset.getEditVersion().setReleaseTime(updateTime);
            theDataset.getEditVersion().setLastUpdateTime(updateTime);
            theDataset.setModificationTime(updateTime);
             ....
```

as opposed to moving all this stuff to the Finalize command? 
This way, if there is a workflow, we are creating this half-published version, that's still a draft, but has the version number, modification dates and other attributes of a published one... Presumably it all needs to be reversed if anything goes wrong with the workflow process. I've already had to move the setPublicationDate() to the Finalize command (since the publication date being non null was the test inside dvObject.isReleased(), so it was confusing the rendering on DatasetPage. 

So far we are inclined to just move all that code to the Finalize command. Would that cause any kind of problem we are not aware of?

**Comment by kcondon** [comment]
At Leonid's request, here is the info for configuring the RSAL Mock:
https://github.com/sbgrid/rsal/tree/master/mocks

Also, to configure Dataverse to use the new workflow you must do the following:
1. Create the workflow:
curl http://localhost:8080/api/admin/workflows -X POST --data-binary @internal-httpSR-workflow.json -H "Content-type: application/json"
2. List available workflows:
curl http://localhost:8080/api/admin/workflows
3. Set the workflow (id) as the default workflow for the appropriate trigger:
curl http://localhost:8080/api/admin/workflows/default/PrePublishDataset -X PUT -d 2
4. Check that the trigger has the appropriate default workflow set:
curl http://localhost:8080/api/admin/workflows/default/PrePublishDataset
5. When finished testing, unset the workflow:
curl -X DELETE http://localhost:8080/api/admin/workflows/default/PrePublishDataset

**Comment by kcondon** [comment]
Issues observed:
Doc:
1. Unset workflow missing $ from triggertype:
DELETE http://$SERVER/api/admin/workflows/default/triggerType
2. Whitelist describes body as list of ip but how separated/formed? Pete says ;
3. Might want a single paragraph or couple sentences describing what workflow is, how configure, current support status.
Behavior:
1. Returning dataset to contributor sends to notifications and emails.
2. Two info messages are listed at top of a dataset submitted for review, 1 green success, submitted for review, 1 brownish yellow saying dataset is now in review.

**Comment by landreev** [comment]
Just pushed in a fix for the 3. (submit for review button needs to go away). 
Thanks @pdurbin for the doc change.
